### PR TITLE
Testing utils

### PR DESCRIPTION
**Pull request by palkan**

### CODE REVIEW

**Comment by palkan** [comment]
@ssnickolay Added testing tools proposal (only docs); please, take a look

File: docs/testing.md [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+    # raises when there are some unspecified associations
+    is_expected.to have_included_associations(:profile)
+      .with_traits(:with_posts)
+    #=> raises RSpec::Expectations::ExpectationNotMetError
+  end
+end
+```
+
+### `have_included_association`
```
Comment (palkan) [inline docs/testing.md]: Rename to `clone_associations`

**Comment by ssnickolay** [comment]
@palkan look good to me!
p/s/ Of course, If you rename `have_included_association` to `clone_associations` as promised ;)

**Comment by palkan** [comment]
@ssnickolay Please, review again; looks like it's ready)

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,11 @@
 # Change log
 
+## master
+
+- Add `Cloner#partial_apply` method. ([@palkan][])
+
+- Add RSpec matchers `clone_association` / `clone_associations`. ([@palkan][])
```
Comment (ssnickolay) [inline CHANGELOG.md]: Can you add note about previous pr https://github.com/palkan/clowne/pull/17 too?)

File: docs/testing.md [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+class UserCloner < Clowne::Cloner
+  class ProfileCloner
+    nullify :rating
+  end
+
+  include_association :profile, clone_with: ProfileCloner
+
+  nullify :email
+
+  finalize do |_, record, name: nil|
```
Comment (ssnickolay) [inline docs/testing.md]: `finalize do |_, record, name: nil, **|`
?

File: docs/testing.md [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+    record.name = name unless name.nil?
+  end
+
+  trait :copy do
+    init_as do |user, target:|
```
Comment (ssnickolay) [inline docs/testing.md]: `init_as do |user, target:, **|`
?

File: spec/clowne/cloner_spec.rb [hunk 5]
```diff
@@ -153,4 +153,62 @@
...
+
+        finalize { |_, record, coef: 2, **| record.age *= coef }
+      end
+    end
+
+    let(:source_class) { Struct.new(:name, :age, :rating) }
+
+    let(:source) { source_class.new('John', 28, 99) }
+
+    specify 'one action', :aggregate_failures do
```
Comment (ssnickolay) [inline spec/clowne/cloner_spec.rb]: You can move `aggregate_failures` to `describe` and remove from all `specify`s

File: lib/clowne/cloner.rb [hunk 6]
```diff
@@ -59,7 +61,11 @@ def call(object, **options)
 
         plan = Clowne::Planner.enhance(plan, Proc.new) if block_given?
 
-        adapter.clone(object, plan, params: options)
+        adapter.clone(object, plan, params: options, only: only)
```
Comment (ssnickolay) [inline lib/clowne/cloner.rb]: I don't really prefer idea with additional `only` key and processing it within a `inject cycle`...How do you think it is possible to filter out (or maybe generate new with option `only`) our plan at this stage and
1. Get ride of `only` key
2. Don't touch the logic of `Adaptes::Base.clone`

?

Also I don't see how you pass this `only` key to nested cloners.

Comment (palkan) [reply lib/clowne/cloner.rb]: Yeah, make sense. What about:

```ruby
plan.filter_declarations(only) unless only.nil?
```

?

Comment (ssnickolay) [reply lib/clowne/cloner.rb]: Sounds good! 👍 
Except that you can "hide" `nil?` condition inside the method
```
def filter_declarations(only)
   return self if only.nil?
end

# and get a clean call ^^
plan.filter_declarations(only)
```

**Comment by palkan** [comment]
@ssnickolay Done

File: lib/clowne/cloner.rb [hunk 7]
```diff
@@ -59,9 +61,15 @@ def call(object, **options)
 
         plan = Clowne::Planner.enhance(plan, Proc.new) if block_given?
 
+        plan = Clowne::Planner.filter_declarations(plan, only) if only
```
Comment (ssnickolay) [inline lib/clowne/cloner.rb]: Unnecessary `if only`  ?

File: lib/clowne/planner.rb [hunk 8]
```diff
@@ -26,6 +26,16 @@ def enhance(plan, block)
         end
       end
 
+      def filter_declarations(plan, only)
+        return if only.nil?
```
Comment (ssnickolay) [inline lib/clowne/planner.rb]: `return plan if only.nil?` ? 🤔

File: CHANGELOG.md [hunk 9]
```diff
@@ -1,5 +1,13 @@
...
+
+- [[#15](https://github.com/palkan/clowne/issues/15)] Add control over nested params. ([@ssnickolay][])
```
Comment (ssnickolay) [inline CHANGELOG.md]: Thx! 👍

**Comment by ssnickolay** [comment]
Awesome 🎉 ! Now we have a whole framework for cloning (:

**Pull request merged**